pyiron_base/project/maintenance.py

```diff
@@ -60,6 +67,75 @@ def get_repository_status():
         return report
 
 
+class UpdateMaintenance:
+    def __init__(self, project):
+        self._project = project
+
+    def base_to_current(self, start_version: str, project=None):
+        """Runs all updates for pyiron_base to reach the current version.
+
+        Args:
+            start_version(str): Version of pyiron_base in the mayor.minor[.patch] format from which to start applying
+                the updates.
+            project(None/project/list/str): The project(s) to be converted from 0.3 to 0.4 ; default: current project
+                One may provide a pyiron Project, a list of pyiron Projects, or a string containing "all" or a valid
+                path.
+                If "all" is provided, pyiron tries to find all projects using the PROJECT_PATHS defined in the
+                configuration.
+        """
+        mayor, minor = start_version.split(".")[0:2]
+        if mayor != 0:
+            raise ValueError("Updates to version >0.x.y is not possible.")
+        if minor < 4:
+            self.base_v0_3_to_v0_4(project)
+
+    def base_v0_3_to_v0_4(self, project=None):
+        """Update hdf files written with pyiron_base-0.3.x to pyiron_base-0.4.x
+
+        pyiron_base<=0.3.9 has a bug that writes all arrays with dtype=object even
+        numeric ones.  As a fix pyiron_base=0.4.0 introduces a conversion when reading
+        such arrays, but does not automatically save them.  This conversion script
+        simply goes over all jobs and rewrites their HDF5 files, since it's read with
+        the correct dtype, this then writes this correct dtype.
+
+        Args:
+            project(None/project/list/str): The project(s) to be converted from 0.3 to 0.4 ; default: current project
+                One may provide a pyiron Project, a list of pyiron Projects, or a string containing "all" or a valid
+                path.
+                If "all" is provided, pyiron tries to find all projects using the PROJECT_PATHS defined in the
+                configuration.
+        """
+        if project is None:
+            projects = [self._project]
+        elif isinstance(project, list):
+            projects = project
+        elif project == "all":
+            projects = [
+                self._project.__class__(path)
+                for path in state.settings.configuration["project_paths"]
+            ]
+        elif isinstance(project, str):
+            if os.path.isdir(project):
+                projects = [self._project.__class__(project)]
+            else:
+                raise ValueError(
+                    f"{project} is a str but neither 'all' nor a directory."
+                )
+        else:
+            projects = [project]
+
+        if len(projects) == 0:
+            raise ValueError(
+                f"Provided project {project} lead to 0 projects to be converted."
+            )
```

Comment:
This can only when providing `project='all'`, right? In that case a warning might be enough.

Reply:
Yes, you are right!